### DISCUSSION TITLE
User Op - op.init called twice

### DISCUSSION THREAD

**Discussion by andrejhronco** [cables-gl/cables_docs#177] [! Welcome - readme first!]
Hi, In my user op, i'm using `op.init` and `op.onLoaded` when logging from each of those functions I see `init` logged twice. 
When getting a port value, the first `init` call shows the port _default_ value and the second shows the _current_ value.
Could be a cause of confusion and potential bugs. 
This is the order in console:

```
op.init - default
op.onLoaded - current
op.init - current
```

**Comment by steam0r** [comment]
thanks for the hint, this really is no problem and would break things if done in a different way. is this causing any issues for you?

**Reply by andrejhronco** [reply]
It was yeah. I was creating an instance of `Tone.Oscillator` which caused an orphaned oscillator that had params that couldn't be changed (there was a mystery oscillator buzzing loudly at the default settings).
I realized I didn't need to use it and refactored and now things are fine.
i'm curious of the cause if you can explain further.
Thanks

**Reply by steam0r** [reply]
hm, hard to explain, has something to do with how cables handles adding and execution of ops. 

in general if you do "init" stuff and then create instances of something (like you seem to be doing) you should always check if that instance is already set, to handle reloading/execution of the op properly. 

so something like this makes sense:

```
let myInstance = null;
op.init = () => {
   if(!myInstance) myInstance = new MyInstance();
}
```

that being said its sometimes an even better idea to give the op an "active/inactive" state via a boolean or a trigger
and then have more control over when things are being initialized:

```
const inActive = op.inBool("Active", false);

let myInstance = null;
inActive.onChange = () => {
    if(inActive.get()) {
        myInstance = new MyInstance();
        op.log("created instance", myInstance);
    }else{
        myInstance = null;
        op.log("cleanup on inactivation", myInstance);
    }
}
```

hope this helps a bit...

**Reply by andrejhronco** [reply]
thanks those are some good approaches to prevent the double init call from being an issue.

**Reply by andrejhronco** [reply]
It would be good to have the documentation updated to note this behavior
